apps/frontend/src/hooks/useSortingAndFiltering.ts

```diff
@@ -0,0 +1,131 @@
+import { useEffect, useMemo, useState } from 'react';
+import SortOrder from '../enums/SortOrder';
+import type FormData from '../types/FormData';
+import { Options } from '../types/SortAndFilter';
+
+/**
+ * This hook is for sorting, filtering, and searching an array of form data.
+ * Currently, you can filter by any number of fields, sort on any field, and searching by most fields
+ * @param list a list of form entries
+ * @param options an object representing the different ways the user wants to sort, filter, and search the data
+ * @returns a piece of state representing the list that is reduced to the user's preferences.
+ */
+const useSortingAndFiltering = (list: FormData[], options: Options) => {
+  const [listState, setListState] = useState(list);
+
+  const reducedData = useMemo(() => reduceData(list, options), [list, options]);
+
+  useEffect(() => {
+    setListState(reducedData);
+  }, [reducedData]);
+
+  return listState;
+};
+
+/**
+ * A function that represents the driver for sorting, searching, and filtering a list of form entries.
+ * @param list A list of form entries
+ * @param options An object that represents every sorting, filtering, and searching option that the user wants
+ * @returns A list of form entries that pass all filtering, sorting, and searching conditions
+ */
+const reduceData = (list: FormData[], options: Options): FormData[] => {
+  const { sorting, filtering, searching } = options;
+
+  let reducedList = [...list];
+
+  if (filtering) {
+    reducedList = filterData(reducedList, { filtering });
+  }
+
+  // TODO: Make searching work with numbers and recursively through all child objects
+  if (searching) {
+    reducedList = searchData(reducedList, { searching });
+  }
+
+  if (sorting) {
+    reducedList = sortData(reducedList, { sorting });
+  }
+
+  return reducedList;
+};
+
+/**
+ * A function that performs filtering on a list of form objects.
+ * @param list A list of form objects
+ * @param filteringObject Options pertaining to filters
+ * @returns A list of filtered form objects
+ */
+const filterData = (
+  list: FormData[],
+  filteringObject: Pick<Options, 'filtering'>
+) => {
+  if (filteringObject && filteringObject.filtering) {
+    const { filters } = filteringObject.filtering;
+    const filteredList = list.filter((formEntry) => {
```

Comment:
I didn't spend much time trying to figure out how to test this but I think this is equivalent and more idiomatic Javascript:
```
    const filteredList = list.filter((formEntry) => {
      return filters.some((filter) => {
        const { field: filterField, value: filterValue } = filter;
        return formEntry[filterField] === filterValue;
      });
    });
```

Reply:
also just to clarify, is the intention for filtering to allow form entries that match at least 1 filter (instead of all filters)? e.g. using `filters.some` instead of `filters.all`?